src/engine/ult.c

```diff
@@ -310,59 +310,54 @@ sched_ult2xs(int xs_type, int tgt_id)
 	switch (xs_type) {
 	case DSS_XS_SELF:
 		return DSS_XS_SELF;
+	case DSS_XS_SYS:
+		return 0;
+	case DSS_XS_SWIM:
+		return 1;
+	case DSS_XS_DRPC:
+		return 2;
 	case DSS_XS_IOFW:
 		if (!dss_helper_pool) {
 			if (dss_tgt_offload_xs_nr > 0)
 				xs_id = DSS_MAIN_XS_ID(tgt_id) + 1;
 			else
-				xs_id = DSS_MAIN_XS_ID((tgt_id + 1) %
-						       dss_tgt_nr);
-			D_ASSERT(xs_id < DSS_XS_NR_TOTAL &&
-				 xs_id >= dss_sys_xs_nr);
-			return xs_id;
+				xs_id = DSS_MAIN_XS_ID((tgt_id + 1) % dss_tgt_nr);
```

Comment:
tgt_id = 7
DSS_MAIN_XS_ID((tgt_id + 1) => (8 * 0 / 8) + 1 + 3 => 4 - this is first xstream (for tgt_id = 0).
So, it will be next xstream in circle. Is this expected other?

Reply:
DSS_MAIN_XS_ID((tgt_id **+ 1**) dss_sys_xs_nr is 3, dss_tgt_nr = 8
so, 8 * (0/8 + 1) + 3 = 11
11 % 8 = 3

Reply:
ah, yes, with this patch, sys_xs_nr is 3,  but 3 does not look right neither. I think it should be itself 10 in this case.

Reply:
@liuxuezhao, @liw Could you comment about this please? From my point of view it's correct but maybe I don't understand the expected logic here.

Reply:
ah, actually my previous calculation is wrong, let's say we have 8 targets, so tgt_id will start from 0 to 7,   then for tgt_id = 7,   

forward_xs_id = main_xs_id ( (7 + 1) % 8) = main_xs_id(0) = 3   does not look right.

So it should be xs_id = DSS_MAIN_XS_ID(tgt_id); ?

Reply:
this if the case that no helper XS, so for IOFW, we either use itself, or use neighbor XS.
Why original code select neighbor XS rather than itself is because, when the code is called, I know myself is on processing IO request and need IO forwarding, now I am processing IO, so likely there is not only one IO (possibly more than one IO for specific dkey), I am busy so likely my neighbor is not busy (both busy seems only in some special multiple dkeys used at same time) can help me do the IO forwarding?
But this is just original intention, you guys think it is not reasonable? that I am processing IO and need IO forwarding, OK, just let myself do it ...

Reply:
Ah, that's what I did not know. I don't have an opinion here and am fine with the existing code. Probably you can add a comment in some PR, because this info looks very valuable. Thanks, Xuezhao.

Reply:
Ok. will add comment.